### PR TITLE
add Helper Encryption for Api Key

### PR DESCRIPTION
**Pull request by Benjamin-Freoua-Alma**

### CODE REVIEW

File: alma/controllers/hook/GetContentHookController.php [hunk 1]
```diff
@@ -92,8 +95,9 @@ public function processConfiguration()
         }
 
         // Down here, we know the provided API keys are correct (at least the one for the chosen API mode)
-        Settings::updateValue('ALMA_LIVE_API_KEY', $liveKey);
-        Settings::updateValue('ALMA_TEST_API_KEY', $testKey);
+        $apiKey = new ApiKeyHelper();
+        $apiKey->setLiveApiKey($liveKey);
+        $apiKey->setTestApiKey($testKey);
```
Comment (Francois-Gomis) [inline alma/controllers/hook/GetContentHookController.php:100]: I think you should add the save responsibility to the controller and remove it from the save helper.

File: alma/lib/Utils/ApiKeyHelper.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+     *
+     * @param string $keyMode
+     * @param string $apiKey
+     *
+     * @return void
+     */
+    public function setLiveApiKey($apiKey)
+    {
+        $encryption = new EncryptionHelper();
+        if ($apiKey != self::OBCUR_VALUE) {
```
Comment (Francois-Gomis) [inline alma/lib/Utils/ApiKeyHelper.php:65]: I think you should remove the save responsibility to the helper and add it to the controller. ;)

File: alma/lib/Utils/ApiKeyHelper.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+     *
+     * @param string $keyMode
+     * @param string $apiKey
+     *
+     * @return void
+     */
+    public function setTestApiKey($apiKey)
+    {
+        $encryption = new EncryptionHelper();
+        if ($apiKey != self::OBCUR_VALUE) {
```
Comment (Francois-Gomis) [inline alma/lib/Utils/ApiKeyHelper.php:81]: I think you should remove the save responsibility to the helper and add it to the controller. ;)

File: alma/lib/Utils/Settings.php [hunk 4]
```diff
@@ -235,7 +236,8 @@ public static function getActiveAPIKey()
      */
     public static function getLiveKey()
     {
-        return self::get('ALMA_LIVE_API_KEY', '');
+        $encryption = new EncryptionHelper();
+        return $encryption->decrypt(self::get(ApiAdminFormBuilder::ALMA_LIVE_API_KEY, ''));
```
Comment (Francois-Gomis) [inline alma/lib/Utils/Settings.php:240]: self::get second parameter is an empty string to change parameter type? null is not good?

File: alma/views/templates/admin/_configure/helpers/form/15/form.tpl [hunk 5]
```diff
@@ -278,6 +278,14 @@
                                                 class="{if isset($input.class)}{$input.class}{/if}"
                                                 value=""
                                                 {if isset($input.autocomplete) && !$input.autocomplete}autocomplete="off"{/if} />
+                                    {elseif $input.type == 'secret'}
+                                        {assign var='value_text' value=$fields_value[$input.name]}
+                                        <input type="password"
+                                            name="{$input.name}"
+                                            size="{$input.size}"
+                                            class="{if isset($input.class)}{$input.class}{/if}"
+                                            value="{if !empty($value_text)}*****{/if}"
```
Comment (Francois-Gomis) [inline alma/views/templates/admin/_configure/helpers/form/15/form.tpl:287]: Can you use ApiHelper OBSCUR const in .TPL file?

File: alma/views/templates/admin/_configure/helpers/form/form.tpl [hunk 6]
```diff
@@ -53,6 +53,14 @@
                 {/strip}
             </div>
         {/foreach}
+    {else if $input.type == 'secret'}
+        {assign var='value_text' value=$fields_value[$input.name]}
+        <input type="password"
+            name="{$input.name}"
+            size="{$input.size}"
+            class="{if isset($input.class)}{$input.class}{/if}"
+            value="{if !empty($value_text)}*****{/if}"
```
Comment (Francois-Gomis) [inline alma/views/templates/admin/_configure/helpers/form/form.tpl:62]: Can you use ApiHelper OBSCUR const in .TPL file?

File: alma/lib/Utils/Settings.php (hunk 4 above)
Comment (Benjamin-Freoua-Alma) [reply alma/lib/Utils/Settings.php:240]: @Francois-Gomis I think it's a same result, I replaced by null :)

File: alma/views/templates/admin/_configure/helpers/form/15/form.tpl (hunk 5 above)
Comment (Benjamin-Freoua-Alma) [reply alma/views/templates/admin/_configure/helpers/form/15/form.tpl:287]: @Francois-Gomis 
I used the placeholder key.
Tell me what do you think ?

File: alma/views/templates/admin/_configure/helpers/form/form.tpl (hunk 6 above)
Comment (Benjamin-Freoua-Alma) [reply alma/views/templates/admin/_configure/helpers/form/form.tpl:62]: @Francois-Gomis 
I used the placeholder key.
Tell me what do you think ?

**Comment by Benjamin-Freoua-Alma** [comment]
@Francois-Gomis 
I added also a condition to see if the key is encrypted.
I forgot about that 😅

**Comment by Benjamin-Freoua-Alma** [comment]
@Francois-Gomis 
I update the responsibility to the setKey.
I made a function for this. I hope it's well-make :)

**Review by Francois-Gomis — APPROVED** [review]

**Comment by Benjamin-Freoua-Alma** [comment]
@Francois-Gomis @ClaireAlma 
I added the encryption at the update module and fixe for Prestashop 1.6/1.5
Commit : 
- add encryption to update
- fix if key no crypted for php old version
- fix Class Rijndael for Prestashop 1.6

File: alma/controllers/hook/GetContentHookController.php [hunk 7]
```diff
@@ -355,13 +373,34 @@ public function processConfiguration()
...
+     * Check if Api key are obscur
+     *
+     * @param string $apiKey
+     * @param string $mode
+     *
+     * @return void|null
+     */
+    private function setKeyIfValueIsNotObscur($apiKey, $mode)
+    {
+        if ($apiKey == ApiKeyHelper::OBCUR_VALUE) {
```
Comment (ClaireAlma) [inline alma/controllers/hook/GetContentHookController.php:386]: === faire de la vrai comparaison, pareil en dessous

File: alma/upgrade/upgrade-2.12.0.php [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+use Alma\PrestaShop\Utils\ApiKeyHelper;
+use Alma\PrestaShop\Utils\Settings;
+
+if (!defined('_PS_VERSION_')) {
+    exit;
+}
+
+function upgrade_module_2_12_0()
+{
+    if (Settings::isFullyConfigured()) {
```
Comment (ClaireAlma) [inline alma/upgrade/upgrade-2.12.0.php:37]: Pourquoi ne retourner que des true et aucun false? Quel est l'interet ?

**Review by ClaireAlma — CHANGES_REQUESTED** [review]

File: alma/controllers/hook/GetContentHookController.php [hunk 9]
```diff
@@ -355,13 +373,34 @@ public function processConfiguration()
         return $this->module->display($this->module->file, 'getContent.tpl');
     }
 
+    /**
+     * Check if Api key are obscur
+     *
+     * @param string $apiKey
+     * @param string $mode
+     *
+     * @return void|null
```
[the remark this answers is not in the document]
Comment (Benjamin-Freoua-Alma) [reply alma/controllers/hook/GetContentHookController.php:382]: Quand elle fait le `setXXXApiKey` elle return rien, du coup je dois mettre que le null ?

File: alma/lib/Forms/AbstractAdminFormBuilder.php [hunk 10]
```diff
@@ -196,6 +198,34 @@ protected function inputTextForm($name, $label, $desc = null, $placeholder = nul
...
+    protected function inputSecretForm($name, $label, $desc = null, $placeholder = ApiKeyHelper::OBCUR_VALUE, $required = false, $lang = false)
+    {
+        $dataInput = [
+            'name' => $name,
+            'label' => $label,
+            'type' => 'secret',
+            'size' => 75,
+            'required' => $required,
+        ];
+
```
[the remark this answers is not in the document]
Comment (Benjamin-Freoua-Alma) [reply alma/lib/Forms/AbstractAdminFormBuilder.php:215]: Mais du coup je dois initier mes $paramName ?

File: alma/upgrade/upgrade-2.12.0.php (hunk 8 above)
Comment (Benjamin-Freoua-Alma) [reply alma/upgrade/upgrade-2.12.0.php:37]: C'est le fichier d'update de module de prestashop, je me suis basé sur les anciens fichier d'update.
J'imagine que c'est pour ne pas bloquer l'update qu'il renvoie true

File: alma/lib/Forms/AbstractAdminFormBuilder.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (ClaireAlma) [reply alma/lib/Forms/AbstractAdminFormBuilder.php:215]: J'ai pas compris on peut en parler si tu veux :)

File: alma/controllers/hook/GetContentHookController.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (ClaireAlma) [reply alma/controllers/hook/GetContentHookController.php:382]: return;

[the remark this answers is not in the document]
Comment (Benjamin-Freoua-Alma) [reply alma/controllers/hook/GetContentHookController.php:382]: @ClaireAlma done

**Review by ClaireAlma — APPROVED** [review]

**Review by Francois-Gomis — APPROVED** [review]

**Pull request merged**